app/models/task.rb

```
@@ -170,6 +170,49 @@ def self.modify_params(params)
     params
   end
 
+  def update_task_type(params)
+    ActiveRecord::Base.transaction do
+      new_branch_task = first_ancestor_of_type.change_type(params)
+      new_child_task = new_branch_task.last_descendant_of_type
+      new_child_task.update!(status: status)
+
+      if assigned_to.is_a?(User) && new_child_task.assigned_to.is_a?(User) &&
+         parent.assigned_to_same_org?(new_child_task.parent)
+        new_child_task.update!(assigned_to: assigned_to)
+      end
+
+      children.active.each { |child| child.update!(parent_id: last_descendant_of_type.id) }
+      first_ancestor_of_type.cancel_descendants
```

Can we add a few code comments around this code explaining why we're doing what we're doing?

---

is `children.active.each { |child| child.update!(parent_id: last_descendant_of_type.id) }` even necessary? Will there ever be any children under the task we're changing the type of?

---

Good question! I don't know. I think it is nice to be a little defensive in this case (also, I'm now realizing this should probably be `children.open` instead of `children.active`). Alternatively we could raise an error here to prevent the task type change from changing and alerting us to the fact that we did end up in this situation.